### ISSUE TITLE
Various edge cases for THREE library variable detection.

### ISSUE THREAD

**Issue by coballast** [coballast/threejs-browserify-conversion-utility#3]
Currently, the code for identifying THREE.js variables only identifies them when they are member expressions. For example, it will pickup Texture in this:
```
THREE.Texture = ...
```

But it won't pick up REVISION in this:
```
THREE = {REVISION: ...}
```

There are also places where properties of THREE are referenced through a variable, which is set to a THREE object based on some piece of code. One example is in MaterialLoader, where this happens:

```
var material = new THREE[ json.type ];
```
In this case, it picks up json.type as a property of THREE, and proper dependencies won't get identified.

**Comment by kumavis** [comment]
We'll need to hone our target here -- we want a `THREE` property access on the left hand side of an assignment statement.

**Comment by coballast** [comment]
Do you mean change the src/ directory to accommodate the way the software finds THREE properties now? Or do you mean change the conversion utility to handle all possible cases?

I think the former would be far easier and faster, although it would be a bit of a manual process.

**Comment by coballast** [comment]
In the scenario where material is being assigned to THREE[json.type] the problem is far more complicated, because in principle the value of json.type could be anything in the three library. Without more information about what can be loaded in json.type, we can't narrow possible dependencies down for MaterialLoader. That is really outside the domain of static analysis. We would need more information that simply isn't available in the library at this point.

**Comment by kumavis** [comment]
so, from my understanding, we still need `Three.thing = thing` in order to produce a three.js build that works exactly the same (thats our goal right? we should be explicit what are goals/restrictions/requirements are, maybe in the README). That also means we don't need to worry about the `var material = new THREE[ json.type ];` case.

I look at this three.js restructuring project as a multiple phase thing. The hardest part is moving browserify. After that we can put pressure on other best practices. The serialization/loader situation presents its own challenges. I briefly mentioned that [here](https://github.com/mrdoob/three.js/issues/4776#issuecomment-42362944). I would not try to tackle that yet.

**Comment by coballast** [comment]
That's the idea, yes. 

The utility performs the processing in several passes.

In the first pass, it goes through files and looks for statements that assign a property of THREE to some value. In otherwords, it looks for "library object declarations", like THREE.Matrix3 = whatever.
It turns these statements into local variables, so it would be changed to var Matrix3 = whatever. It then makes a record in a data structure indicating the particular file, and what was declared in the particular file. It also records THREE variables what were simply mentioned, but not declared. This is how it infers dependencies in future passes. <a href="https://github.com/coballast/threejs-browserify-conversion-utility/blob/master/asts-dependencies.js">Here is the code</a>

It excludes the root level Three.js file and treats it as special. All properties that were assigned directly to THREE in that file that everything else needs to work are left alone everywhere so as to preserve the api to the external world.

At the end of all of this magic, the utility goes into the THREE.js file and appends a massive list of require statements of the form "THREE.whatever = require('relative-path-to-whatever')" followed by a "module.exports = THREE".

So my primary problem is that if we are given a variable that indexes a property of THREE rather than an explicit property, then there is no way to infer dependencies. That's the crux of it.

**Comment by coballast** [comment]
A quick hack that would fix the problem is to simply include every possible external dependency in files like that. Browserify might optimize that stuff away in the final build, and it could be eliminated by hand by a human if/when this makes it into the repo.

**Comment by kumavis** [comment]
Ah, thanks that is a tricky problem.

Is this common? if its only in a couple files, maybe we can just provide a list of all modules to include `materials/*.js`.

**Comment by coballast** [comment]
I can detect the files that this happens in pretty easily. I can generate a  list, and my intuition says that it will be a really short list.

**Comment by coballast** [comment]
I was right. Here are the files:

loaders/Loader.js
Reason:
Indexes THREE on blending mode and other things

loaders/MaterialLoader.js
Reason:
Indexes THREE on material type as mentioned above.

Probably similar things happen in the other files. I'm too lazy to read through them:
loaders/ObjectLoader.js
renderers/WebGLRenderer.js

**Comment by kumavis** [comment]
The `Loaders` make sense, whatever WebGLRenderer is doing should likely be inverted (that code should be put into the respective classes instead of the renderer), but thats out of scope for this transform.

**Comment by coballast** [comment]
I think we can at least remedy the loaders. Maybe do as you suggested, and have another pass that takes a structure that defines files and "supplemental" dependencies that can't be found in the code...

**Comment by coballast** [comment]
I am not sure how to transform the loading code though. The only way I know of to make it instantiate a class dynamically based on a variable without indexing some parent object like window or THREE is to use eval. Ughhh...

**Comment by kumavis** [comment]
Maybe we need `Three.js` to be just `module.exports = {}` and then we have a different entrypoint `index.js` that requires THREE and loads all the classes onto it. That way we can get a reference to `Three.js` thats not circular... maybe?

**Comment by coballast** [comment]
I think this has been effectively fixed for the same reasons as #9
